File: chapter4.mq/src/main/java/com/brownfield/pss/client/MqApp.java
package com.brownfield.pss.client;

import com.github.dockerjava.api.command.CreateContainerCmd;
import com.github.dockerjava.api.model.ExposedPort;
import com.github.dockerjava.api.model.PortBinding;
import com.github.dockerjava.api.model.Ports;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.testcontainers.containers.FixedHostPortGenericContainer;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import java.time.Duration;
import java.util.function.Consumer;

import static java.time.temporal.ChronoUnit.SECONDS;


@SpringBootApplication
public class MqApp {
    private static final Logger logger = LoggerFactory.getLogger(MqApp.class);
    private static final int containerExposedPort = 5672;

    private static GenericContainer rabbitMq = new FixedHostPortGenericContainer("rabbitmq:3-management")
            .withFixedExposedPort(containerExposedPort, containerExposedPort)
            .withFixedExposedPort(15672, 15672)
            .waitingFor(new LogMessageWaitStrategy().withRegEx(".*Server startup complete.*\\s"))
            .withStartupTimeout(Duration.of(10, SECONDS));

    public static void main(String[] args) {
        SpringApplication.run(MqApp.class, args);
    }

    @PostConstruct
    public void start() {
        rabbitMq.start();
    }

    @PreDestroy
    public void tearDown() {
        rabbitMq.stop();
    }
}

File: chapter4.mq/src/main/java/com/brownfield/pss/client/MqApp.kt
package com.brownfield.pss.client

import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.testcontainers.containers.FixedHostPortGenericContainer
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy
import java.time.Duration
import java.time.temporal.ChronoUnit.SECONDS
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy

class KGenericContainer(imageName: String) : FixedHostPortGenericContainer<KGenericContainer>(imageName)

private val rabbitMq = KGenericContainer("rabbitmq:3-management")
    .withFixedExposedPort(5672, 5672)
    .withFixedExposedPort(15672, 15672)
    .waitingFor(LogMessageWaitStrategy().withRegEx(".*Server startup complete.*\\s"))
    .withStartupTimeout(Duration.of(10, SECONDS))

@SpringBootApplication
class MqApp {
    @PostConstruct
    fun start() = rabbitMq.start()

    @PreDestroy
    fun tearDown() = rabbitMq.stop()
}

fun main(args: Array<String>) {
    SpringApplication.run(MqApp::class.java, *args)
}